File: android/employer/app/src/main/java/com/workforcex/employer/ui/employer/JobCreateEditActivity.java
package com.workforcex.employer.ui.employer;

import android.os.Bundle;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.ArrayAdapter;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import com.google.android.material.chip.Chip;
import com.workforcex.employer.api.JobRequest;
import com.workforcex.employer.api.JobResponse;
import com.workforcex.employer.api.RetrofitClient;
import com.workforcex.employer.api.Skill;
import com.workforcex.employer.databinding.ActivityJobCreateEditBinding;
import com.workforcex.employer.utils.TokenManager;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class JobCreateEditActivity extends AppCompatActivity {

    private ActivityJobCreateEditBinding binding;
    private TokenManager tokenManager;
    private String jobId;

    private final Set<String> selectedSkills = new LinkedHashSet<>();
    private List<String> allSkills = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityJobCreateEditBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        tokenManager = new TokenManager(this);

        jobId = getIntent().getStringExtra("jobId");
        boolean isEditing = jobId != null;
        setTitle(isEditing ? "Edit Job" : "Create Job");

        fetchSkillsAndSetupAutocomplete();

        if (isEditing) {
            binding.etTitle.setText(getIntent().getStringExtra("jobTitle"));
            addSkillsFromCsv(getIntent().getStringExtra("jobSkills"));
            int exp = getIntent().getIntExtra("jobExperience", 0);
            if (exp > 0) binding.etExperience.setText(String.valueOf(exp));
            binding.etLocation.setText(getIntent().getStringExtra("jobLocation"));
            double salMin = getIntent().getDoubleExtra("jobSalaryMin", 0);
            double salMax = getIntent().getDoubleExtra("jobSalaryMax", 0);
            if (salMin > 0) binding.etSalaryMin.setText(String.valueOf((int) salMin));
            if (salMax > 0) binding.etSalaryMax.setText(String.valueOf((int) salMax));
            int positions = getIntent().getIntExtra("jobOpenPositions", 0);
            if (positions > 0) binding.etOpenPositions.setText(String.valueOf(positions));
            binding.etDescription.setText(getIntent().getStringExtra("jobDescription"));
        }

        binding.btnSave.setText(isEditing ? "Update Job" : "Post Job");
        binding.btnSave.setOnClickListener(v -> saveJob());
    }

    private void fetchSkillsAndSetupAutocomplete() {
        RetrofitClient.get().getSkills().enqueue(new Callback<List<Skill>>() {
            @Override
            public void onResponse(Call<List<Skill>> call, Response<List<Skill>> response) {
                if (response.isSuccessful() && response.body() != null) {
                    allSkills = response.body().stream().map(s -> s.name).collect(Collectors.toList());
                    setupSkillsAutocomplete();
                }
            }
            @Override
            public void onFailure(Call<List<Skill>> call, Throwable t) {}
        });
    }

    private void setupSkillsAutocomplete() {
        ArrayAdapter<String> adapter = new ArrayAdapter<>(
                this, android.R.layout.simple_dropdown_item_1line, allSkills);
        binding.acSkills.setAdapter(adapter);

        binding.acSkills.setOnItemClickListener((parent, view, position, id) -> {
            String skill = (String) parent.getItemAtPosition(position);
            addSkillChip(skill);
            binding.acSkills.setText("");
        });

        binding.acSkills.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                String typed = binding.acSkills.getText().toString().trim().toLowerCase();
                if (!typed.isEmpty()) {
                    addSkillChip(typed);
                    binding.acSkills.setText("");
                }
                return true;
            }
            return false;
        });
    }

    private void addSkillChip(String skill) {
        if (selectedSkills.size() >= 5) {
            Toast.makeText(this, "You can add a maximum of 5 skills", Toast.LENGTH_SHORT).show();
            return;
        }
        String clean = skill.trim().toLowerCase();
        if (clean.isEmpty() || selectedSkills.contains(clean)) return;

        selectedSkills.add(clean);

        Chip chip = new Chip(this);
        chip.setText(clean);
        chip.setCloseIconVisible(true);
        chip.setOnCloseIconClickListener(v -> {
            selectedSkills.remove(clean);
            binding.chipGroupSkills.removeView(chip);
        });
        binding.chipGroupSkills.addView(chip);
    }

    private void addSkillsFromCsv(String csv) {
        if (csv == null || csv.isBlank()) return;
        for (String skill : csv.split(",")) {
            addSkillChip(skill.trim());
        }
    }

    private String getSkillsCsv() {
        return String.join(",", selectedSkills);
    }

    private void saveJob() {
        String title = binding.etTitle.getText().toString().trim();
        if (title.isEmpty()) { binding.etTitle.setError("Title is required"); return; }

        JobRequest request = new JobRequest();
        request.title = title;
        request.skillsRequired = getSkillsCsv();
        request.location = binding.etLocation.getText().toString().trim();
        request.description = binding.etDescription.getText().toString().trim();

        String expStr = binding.etExperience.getText().toString().trim();
        request.experienceRequired = expStr.isEmpty() ? null : Integer.parseInt(expStr);

        String salMinStr = binding.etSalaryMin.getText().toString().trim();
        request.salaryMin = salMinStr.isEmpty() ? null : Double.parseDouble(salMinStr);

        String salMaxStr = binding.etSalaryMax.getText().toString().trim();
        request.salaryMax = salMaxStr.isEmpty() ? null : Double.parseDouble(salMaxStr);

        String posStr = binding.etOpenPositions.getText().toString().trim();
        request.openPositions = posStr.isEmpty() ? null : Integer.parseInt(posStr);

        setLoading(true);

        Callback<JobResponse> callback = new Callback<JobResponse>() {
            @Override
            public void onResponse(Call<JobResponse> call, Response<JobResponse> response) {
                setLoading(false);
                if (response.isSuccessful()) {
                    Toast.makeText(JobCreateEditActivity.this,
                            jobId != null ? "Job updated!" : "Job posted!", Toast.LENGTH_SHORT).show();
                    finish();
                } else {
                    Toast.makeText(JobCreateEditActivity.this, "Failed to save job", Toast.LENGTH_SHORT).show();
                }
            }
            @Override
            public void onFailure(Call<JobResponse> call, Throwable t) {
                setLoading(false);
                Toast.makeText(JobCreateEditActivity.this, "Network error", Toast.LENGTH_SHORT).show();
            }
        };

        if (jobId != null) {
            RetrofitClient.get().updateJob(tokenManager.getBearerToken(), jobId, request).enqueue(callback);
        } else {
            RetrofitClient.get().createJob(tokenManager.getBearerToken(), request).enqueue(callback);
        }
    }

    private void setLoading(boolean loading) {
        binding.progressBar.setVisibility(loading ? View.VISIBLE : View.GONE);
        binding.btnSave.setEnabled(!loading);
    }
}

File: android/employer/app/src/main/java/com/workforcex/employer/ui/employer/JobCreateEditActivity.kt
package com.workforcex.employer.ui.employer

import android.os.Bundle
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.chip.Chip
import com.workforcex.shared_employer.models.JobRequest
import com.workforcex.shared_employer.models.JobResponse
import com.workforcex.shared_employer.RetrofitClient
import com.workforcex.shared_employer.models.Skill
import com.workforcex.employer.databinding.ActivityJobCreateEditBinding
import com.workforcex.employer.utils.TokenManager
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.stream.Collectors

class JobCreateEditActivity : AppCompatActivity() {

    private lateinit var binding: ActivityJobCreateEditBinding
    private lateinit var tokenManager: TokenManager
    private var jobId: String? = null

    private val selectedSkills = LinkedHashSet<String>()
    private var allSkills: List<String> = ArrayList()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityJobCreateEditBinding.inflate(layoutInflater)
        setContentView(binding.root)
        tokenManager = TokenManager(this)

        jobId = intent.getStringExtra("jobId")
        val isEditing = jobId != null
        title = if (isEditing) "Edit Job" else "Create Job"

        fetchSkillsAndSetupAutocomplete()

        if (isEditing) {
            binding.etTitle.setText(intent.getStringExtra("jobTitle"))
            addSkillsFromCsv(intent.getStringExtra("jobSkills"))
            val exp = intent.getIntExtra("jobExperience", 0)
            if (exp > 0) binding.etExperience.setText(exp.toString())
            binding.etLocation.setText(intent.getStringExtra("jobLocation"))
            val salMin = intent.getDoubleExtra("jobSalaryMin", 0.0)
            val salMax = intent.getDoubleExtra("jobSalaryMax", 0.0)
            if (salMin > 0) binding.etSalaryMin.setText(salMin.toInt().toString())
            if (salMax > 0) binding.etSalaryMax.setText(salMax.toInt().toString())
            val positions = intent.getIntExtra("jobOpenPositions", 0)
            if (positions > 0) binding.etOpenPositions.setText(positions.toString())
            binding.etDescription.setText(intent.getStringExtra("jobDescription"))
        }

        binding.btnSave.text = if (isEditing) "Update Job" else "Post Job"
        binding.btnSave.setOnClickListener { saveJob() }
    }

    private fun fetchSkillsAndSetupAutocomplete() {
        RetrofitClient.get().getSkills().enqueue(object : Callback<List<Skill>> {
            override fun onResponse(call: Call<List<Skill>>, response: Response<List<Skill>>) {
                if (response.isSuccessful && response.body() != null) {
                    allSkills = response.body()!!.stream().map { s: Skill -> s.name }.collect(Collectors.toList())
                    setupSkillsAutocomplete()
                }
            }

            override fun onFailure(call: Call<List<Skill>>, t: Throwable) {}
        })
    }

    private fun setupSkillsAutocomplete() {
        val adapter = ArrayAdapter(
            this, android.R.layout.simple_dropdown_item_1line, allSkills
        )
        binding.acSkills.setAdapter(adapter)

        binding.acSkills.setOnItemClickListener { parent, _, position, _ ->
            val skill = parent.getItemAtPosition(position) as String
            addSkillChip(skill)
            binding.acSkills.setText("")
        }

        binding.acSkills.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                val typed = binding.acSkills.text.toString().trim().lowercase()
                if (typed.isNotEmpty()) {
                    addSkillChip(typed)
                    binding.acSkills.setText("")
                }
                return@setOnEditorActionListener true
            }
            false
        }
    }

    private fun addSkillChip(skill: String) {
        if (selectedSkills.size >= 5) {
            Toast.makeText(this, "You can add a maximum of 5 skills", Toast.LENGTH_SHORT).show()
            return
        }
        val clean = skill.trim().lowercase()
        if (clean.isEmpty() || selectedSkills.contains(clean)) return

        selectedSkills.add(clean)

        val chip = Chip(this)
        chip.text = clean
        chip.isCloseIconVisible = true
        chip.setOnCloseIconClickListener {
            selectedSkills.remove(clean)
            binding.chipGroupSkills.removeView(chip)
        }
        binding.chipGroupSkills.addView(chip)
    }

    private fun addSkillsFromCsv(csv: String?) {
        if (csv.isNullOrBlank()) return
        for (skill in csv.split(",")) {
            addSkillChip(skill.trim())
        }
    }

    private fun getSkillsCsv(): String {
        return selectedSkills.joinToString(",")
    }

    private fun saveJob() {
        val title = binding.etTitle.text.toString().trim()
        if (title.isEmpty()) {
            binding.etTitle.error = "Title is required"
            return
        }

        val request = JobRequest(
            title = title,
            skillsRequired = getSkillsCsv(),
            experienceRequired = binding.etExperience.text.toString().trim().toIntOrNull() ?: 0,
            location = binding.etLocation.text.toString().trim(),
            salaryMin = binding.etSalaryMin.text.toString().trim().toDoubleOrNull() ?: 0.0,
            salaryMax = binding.etSalaryMax.text.toString().trim().toDoubleOrNull() ?: 0.0,
            openPositions = binding.etOpenPositions.text.toString().trim().toIntOrNull() ?: 0,
            description = binding.etDescription.text.toString().trim()
        )

        setLoading(true)

        val callback = object : Callback<JobResponse> {
            override fun onResponse(call: Call<JobResponse>, response: Response<JobResponse>) {
                setLoading(false)
                if (response.isSuccessful) {
                    Toast.makeText(
                        this@JobCreateEditActivity,
                        if (jobId != null) "Job updated!" else "Job posted!", Toast.LENGTH_SHORT
                    ).show()
                    finish()
                } else {
                    Toast.makeText(this@JobCreateEditActivity, "Failed to save job", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<JobResponse>, t: Throwable) {
                setLoading(false)
                Toast.makeText(this@JobCreateEditActivity, "Network error", Toast.LENGTH_SHORT).show()
            }
        }

        if (jobId != null) {
            RetrofitClient.get().updateJob(tokenManager.getBearerToken(), jobId!!, request).enqueue(callback)
        } else {
            RetrofitClient.get().createJob(tokenManager.getBearerToken(), request).enqueue(callback)
        }
    }

    private fun setLoading(loading: Boolean) {
        binding.progressBar.visibility = if (loading) View.VISIBLE else View.GONE
        binding.btnSave.isEnabled = !loading
    }
}